pkg/auth/handler/ssohandler/auth_handler.go

```diff
@@ -0,0 +1,398 @@
+}
+
+// AuthHandler decodes code response and fetch access token from provider.
+//
+// curl http://localhost:3000/sso/<provider>/auth_handler?code=<code>&state=<state>
+//
+// For ux_mode is 'ios' or 'android',
+// it creates a 302 response, and Location points to:
+// myapp://user.skygear.io/sso/{provider}/auth_handler?result=
+//
+// Fox ux_mode is 'web_redirect',
+// it creates a 302 response, and Location points to: sso_callback_url
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	Provider                sso.Provider                `dependency:"SSOProvider"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderName            string
+}
+
+func (h AuthHandler) WithTx() bool {
+	return true
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) Handler() http.Handler {
+	// reference from APIHandlerToHandler
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		var respErr skyerr.Error
+
+		if h.Provider == nil {
+			respErr = skyerr.NewInvalidArgument("Provider is not supported", []string{h.ProviderName})
+			return
+		}
+
+		defer func() {
+			if respErr != nil {
+				response := handler.APIResponse{
+					Err: respErr,
+				}
+				handler.WriteResponse(rw, response)
+			}
+		}()
+
+		payload, err := h.DecodeRequest(r)
+		if err != nil {
+			respErr = skyerr.MakeError(err)
+			return
+		}
+
+		if err := payload.Validate(); err != nil {
+			respErr = skyerr.MakeError(err)
+			return
+		}
+
+		if h.WithTx() {
+			// assume txContext != nil if apiHandler.WithTx() is true
+			if err := h.TxContext.BeginTx(); err != nil {
+				panic(err)
+			}
+
+			defer func() {
+				if h.TxContext.HasTx() {
+					h.TxContext.RollbackTx()
+				}
+			}()
+		}
+
+		reqPayload := payload.(AuthRequestPayload)
+		oauthAuthInfo, err := h.getAuthInfo(reqPayload)
+		if err != nil {
+			respErr = skyerr.MakeError(err)
+			return
+		}
+		authResp, err := h.getAuthResp(oauthAuthInfo)
+		if h.TxContext != nil {
+			h.TxContext.CommitTx()
+		}
+
+		// handle callback url by ux_mode
+		tConfig := config.GetTenantConfig(r)
```

Comment:
Handler should not read the tenant config, the design of dependency injection is to handle the multi-tenant logic, so handler doesn't need to have knowledge of multi tenant thing. Maybe we should have a sso dependency that hold the SSO setting and help do the callback urls validation.